fix: ensure cell selections are always rectangular

liwenka1 opened the pull request:
## Problem

When selecting table cells with `colspan` or `rowspan` attributes, the current implementation allows non-rectangular selections (T-shaped, L-shaped, etc.). This behavior is inconsistent with mature table editors and can lead to unexpected results.

### Example Issue

When selecting from cell A to cell C in the following table:

```
| A | B (rowspan=2) | C |
| D | B             | E |
```

**Before Fix:** The selection would be T-shaped, only including cells A, B, and C.

**After Fix:** The selection automatically expands to include all cells (A, B, C, D, E), forming a complete rectangle.

## Before Fix

![20251225174815_rec_](https://github.com/user-attachments/assets/3e9f8110-2364-477b-a014-34179aa5c648)

_The selection creates a non-rectangular shape when crossing cells with rowspan/colspan._

## After Fix

![20251225174852_rec_](https://github.com/user-attachments/assets/ca0cc1c8-7af5-4eb1-be62-8ce3d0a18464)

_The selection automatically expands to ensure a rectangular shape, including all affected cells._

## Industry Standards

This fix aligns with the behavior of professional table editors:

### Feishu (Lark) Table Selection
![20251225174706_rec_](https://github.com/user-attachments/assets/518f9eec-c862-4d91-943a-6f747c2953aa)

_Feishu enforces rectangular selections in tables._

### Microsoft Office Table Selection
![img_v3_02ta_1c853afd-ed94-4df9-9dcd-c2283e9422hu](https://github.com/user-attachments/assets/e06247d3-410c-41c9-92c4-a452b42eb242)

_Microsoft Word/Excel also maintains rectangular cell selections._

## Implementation Details

### Core Changes

**Modified `TableMap.rectBetween()` in `src/tablemap.ts`:**
- Added automatic rectangle expansion logic
- Iteratively checks all cells within the current rectangle
- If any cell spans beyond the rectangle boundaries (via `colspan` or `rowspan`), the rectangle expands to include it
- Continues until the rectangle is stable (no more expansion needed)

### Algorithm

```typescript
1. Calculate initial rectangle between anchor and head cells
2. While rectangle is expanding:
   a. For each cell in current rectangle:
      - Get the cell's actual boundaries (considering colspan/rowspan)
      - If cell extends beyond rectangle, expand rectangle to include it
   b. If any expansion occurred, repeat step 2
3. Return the final rectangular selection
```

### Test Coverage

- ✅ Added `test/cellselection-rect.test.ts` with specific tests for rectangular constraints
- ✅ Updated `test/tablemap.test.ts` to reflect new `rectBetween` behavior
- ✅ Updated `test/cellselection.test.ts` - selections now expand instead of cutting off cells
- ✅ Updated `test/commands.test.ts` - commands now operate on expanded rectangular selections

All 181 tests pass ✅

## Benefits

1. **Consistent UX:** Matches user expectations from other table editors (Feishu, Office, Google Docs)
2. **Predictable Behavior:** No more surprising T-shaped or L-shaped selections
3. **Correct Operations:** Table operations (merge, delete, copy) now work on complete rectangular regions
4. **Better Accessibility:** Users can select complete table regions without worrying about internal cell structure

## Breaking Changes

⚠️ **Behavior Change:** Cell selections that previously created non-rectangular shapes will now automatically expand to rectangles. This could affect:
- Custom code that relies on non-rectangular selections
- Tests that expect partial cell selections across spanning cells

However, this change makes the behavior more intuitive and consistent with industry standards.

## Migration Guide

If you have code that depends on the old behavior:

**Before:**
```typescript
// Selection from A to C would only include A, B, C
const selection = new CellSelection($anchorCell, $headCell);
```

**After:**
```typescript
// Selection from A to C now includes A, B, C, D, E (complete rectangle)
const selection = new CellSelection($anchorCell, $headCell);
// The selection is automatically expanded to be rectangular
```

No API changes are required - the expansion happens automatically in the `CellSelection` constructor.

## Related Issues

Fixes issue where table selections could create non-rectangular shapes when cells have colspan or rowspan attributes.

---

**Checklist:**
- [x] All tests pass
- [x] Code follows project style guidelines (Prettier + ESLint)
- [x] New tests added for the feature
- [x] Behavior aligns with industry standards (Feishu, Office)

liwenka1 committed:
fix: ensure cell selections are always rectangular

- Modified TableMap.rectBetween to automatically expand selection rectangles
- Selections now include all cells that span across the boundary
- Prevents T-shaped or L-shaped selections when cells have colspan/rowspan
- Added tests to verify rectangular selection constraint
- Updated existing tests to reflect the new behavior

This change aligns with the behavior of mature table editors like Feishu
and Microsoft Office, where table selections are always rectangular.

ocavue commented on test/cellselection-rect.test.ts:
```diff
@@ -0,0 +1,89 @@
+    // | A  | B (rowspan=2) | C  |
+    // | D  | B             | E  |
+    const tableNode = table(
+      tr(
+        td(p('A')), // pos 1
+        td({ rowspan: 2 }, p('B')), // pos 6
+        td(p('C')), // pos 11
+      ),
```
It's unclear what `pos 1` means here. I recommend using the comment shown [here](https://github.com/ProseMirror/prosemirror-tables/blob/v1.8.5/test/cellselection.test.ts#L35-L37) in the test file.

ocavue commented on test/cellselection-rect.test.ts:
```diff
@@ -0,0 +1,89 @@
+
+    console.log('Map:', map.map);
+    console.log('Rect:', rect);
+    console.log('Cells in rect:', map.cellsInRect(rect));
```
Let's remove these `console.log` in the test file

ocavue commented on test/cellselection-rect.test.ts:
```diff
@@ -0,0 +1,89 @@
+
+    console.log('Map:', map.map);
+    console.log('Rect:', rect);
+    console.log('Cells in rect:', map.cellsInRect(rect));
```
Let's remove these `console.log` in the test file

ocavue commented on src/tablemap.ts:
```diff
@@ -177,12 +177,48 @@ export class TableMap {
+          if (cellRect.bottom > rect.bottom) {
+            rect.bottom = cellRect.bottom;
+            expanded = true;
+          }
```
If I remove all four `expanded = true` here, I can still pass all tests. Could you construct a test case that will fail once `expanded = true` lines are removed?


You can use the following case as an example: 


https://github.com/user-attachments/assets/430c21ba-2538-4626-9dc8-890f0979fa51


https://github.com/user-attachments/assets/139604c0-5e70-480f-89e4-00b6b539bdab

ocavue commented on src/tablemap.ts:
```diff
@@ -177,12 +177,48 @@ export class TableMap {
+        for (let col = rect.left; col < rect.right; col++) {
+          const index = row * this.width + col;
+          const cellPos = this.map[index];
+          const cellRect = this.findCell(cellPos);
```
I believe there is room for performance improvement:

1. We don't need to check _all_ cells in the current rectangle. We only need to check the cells at the four edges. This can reduce the check numbers from $\text{width} \times \text{height}$ to $O(\text{width} + \text{height})$.
2. `this.findCell()` call is slow because it's an $O(\text{width} \times \text{height})$ operation. We should avoid unnecessary `this.findCell()` calls. One method is to use a `seen` object to cache all checked points like [this example](https://github.com/ProseMirror/prosemirror-tables/blob/42851d626a58a7c0faf43be17af9bb5dfe2da888/src/tablemap.ts#L350-L353).

ocavue commented:
This is a great improvement! I left some review comments.

liwenka1 committed:
perf: optimize rectBetween with edge-only checking and caching

- Optimize cell checking from O(width×height) to O(width+height) per iteration
- Add seen cache to avoid redundant findCell() calls
- Update position comments to use standard /* pos */ format
- Remove console.log statements from tests
- Add test case for complex rowspan/colspan expansion validation

Performance improvement: For a 10×10 table, reduced from 100 cells checked
to 40 cells per iteration (60% reduction). The optimization maintains
correctness by checking only edge cells, as interior cells cannot cause
rectangle expansion.

liwenka1 commented:
Thank you @ocavue  for the thorough review and valuable suggestions! I've addressed all the feedback points

RatoX commented:
Quick context: This wasn't a bug but a feature.

forcing a rectangular selection will make some UX behaviours quite difficult to implement like: "change the background color for an entire row"

ocavue commented:
> Quick context: This wasn't a bug but a feature.
> 
> forcing a rectangular selection will make some UX behaviours quite difficult to implement like: "change the background color for an entire row"



That's a good point. I just checked Excel's behavior, and it does provide different behaviors:

1. If I hold the mouse and drag a selection, it always creates a rectangular shape. 
2. If I click the column and row header, I can only select "part of" a merged cell thus I can "change the background color for an entire row".



https://github.com/user-attachments/assets/7594f567-538d-4846-b122-c37d83162f29

ocavue commented:
Perhaps we can add a new option to `new CellSelection($anchorCell, $headCell)` to indicate whether we want to force the selection to be rectangular. When creating row/column selection programmatically, we don't force it. When creating a selection via the mouse move event, we force it.

liwenka1 committed:
feat: add forceRectangular option to CellSelection

liwenka1 commented:
Thanks for the valuable feedback! I've implemented the `forceRectangular` option as suggested.

**Changes:**
- Added `CellSelectionOptions` interface with `forceRectangular?: boolean`
- `CellSelection` constructor now accepts an optional third parameter
- `TableMap.rectBetween()` accepts `forceRectangular` parameter (default: `false`)
- Mouse drag and Shift+Arrow selections use `forceRectangular: true`
- `rowSelection()` and `colSelection()` use default `false`, allowing partial merged cell selection

**Behavior:**

| Selection Method | forceRectangular | Can select partial merged cells |
|-----------------|------------------|--------------------------------|
| Mouse drag | true | No |
| Shift+Arrow | true | No |
| `rowSelection()` | false | Yes |
| `colSelection()` | false | Yes |
| `new CellSelection()` | false (default) | Yes |

This matches Excel's behavior where drag selections are rectangular, but row/column header selections can include partial merged cells.

Please review when you have time. Thanks!

liwenka1 committed:
fix: persist forceRectangular in cell selections
